packages/query-core/src/hydration.ts

```diff
@@ -82,6 +82,9 @@ function dehydrateQuery(
     queryHash: query.queryHash,
     ...(query.state.status === 'pending' && {
       promise: query.promise?.then(serializeData).catch((error) => {
+        if ('digest' in error && error.digest === 'DYNAMIC_SERVER_USAGE') {
+          return Promise.reject(error)
+        }
```

Comment:
i'm assuming we don't want these codes to leak into query-core, so should the `defaultQueryOptions` take some `onPrefetchError`/`onPromise...` or similar prop that let's you specify these errors you want to be re-thrown. I'm guessing you might want `notFound()`, `redirect()` errors to have a similar behavior

Reply:
Added a `shouldRedactError` option that let's you control whether or not the error should be redacted on dehydration: https://github.com/TanStack/query/pull/8383/commits/75849289ba1bf90bf3a24cf31c0c90d3f083e54b

Reply:
we’d need to document this too please

Reply:
done